### ISSUE TITLE
Slack to IRC bridge does not work for the #tor-meeting channel

### ISSUE THREAD

**Issue by hellais** [ooni/backend#298]
The last messages that was relayed from IRC were on January 7th 2020:
```
pili 17:02
#endmeetig
17:02
ugh
17:02
there we go... ;P

sysrqb 17:20
m/win 3 
```

and then silence

According to anarcat:
```
15:27 <+anarcat> hellais: all i can see in my logs is this: 2019-12-26 21:05:11 -!- mode/#tor-meeting [+R] by arma2
```

In theory the slacktopus is configured to authenticate to the NickServ, see: https://github.com/ooni/sysadmin/blob/003d50efcb633b8416d2c3ce4896461362750c5b/ansible/host_vars/mia-slack.ooni.nu/vars.yml#L42, but maybe something is not working there.

We don't see anything unusual from the logs of docker, but we may have to enable more debugging info.

FWIW this only happens with the #tor-meeting channel and not with the other IRC bridged channels.

**Comment by anarcat** [comment]
more info:

```
11:07:16 <+anarcat> this is what i got from the bot in my backlog: 
11:07:16 <+anarcat> 2019-12-31 06:23:57 <slacktopus> <neal> Where is that bot connecting from?
11:07:16 <+anarcat> 2020-01-08 02:15:36 -!- slacktopus [~slacktopu@37.218.241.90] has quit [Remote host closed the connection]
11:11:52 <+anarcat> hellais: before arma2 did the change, there were some spammers in this channel
```

so it probably died on the 8th and never was able to return. i would also suspect the nickserv interaction is failing. looking at that YAML file, i would advise using a client  cert to authentify with the server instead of talking to nickserv, which is more error prone (because of race conditions). It could very well be that you try to join the channel before you're authenticated. An alternative is to send the nickserv password as a server password, which also works.

See https://oftc.net/NickServ/CertFP/ for how to setup a client cert.

**Comment by hellais** [comment]
I spent a bit of time debugging this and didn't really come up with a solution.

I upgraded our docker image of the slack-irc bridge to the latest version and turned on debug level logging on it.
This allowed we to see that the root cause is in fact an unregistered nick:
```
[winston] Attempt to write logs with no transports {"prefix":"reflection.oftc.net","server":"reflection.oftc.net","command":"rpl_welcome","rawCommand":"001","commandType":"reply","args":["slacktopus","Welcome to the OFTC Internet Relay Chat Network slacktopus"],"level":"debug","message":"Registered event: "}
[winston] Attempt to write logs with no transports {"prefix":"reflection.oftc.net","server":"reflection.oftc.net","command":"477","rawCommand":"477","commandType":"error","args":["slacktopus","#tor-meeting","Cannot join channel (Need to be identified and verified to join this channel, '/msg NickServ help' to learn how to register and verify.)"],"level":"error","message":"Received error event from IRC"}
```

I tried using SASL based authentication to get rid of the race in logging in, but I could not get it work with OFTC and I kept getting this error:
```
21 Jan 18:34:43 - ERROR: {
  prefix: 'graviton.oftc.net',
  server: 'graviton.oftc.net',
  command: 'err_notregistered',
  rawCommand: '451',
  commandType: 'error',
  args: [ '*', 'slacktopus', 'Register first.' ]
}
```

I didn't do enough debugging to understand why it's not working though. It's also odd that even after some time the slacktopus nick appears to not be registered as if the autoSendCommand arguments are getting ignored.

In any case more debugging is needed and this is not as easy as I had thought it would be.

If we could drop the +R from the #tor-meeting channel that would be ideal, but otherwise we will have to debug it more...

**Comment by anarcat** [comment]
> I didn't do enough debugging to understand why it's not working though. It's also odd that even after some time the slacktopus nick appears to not be registered as if the autoSendCommand arguments are getting ignored.

Have you tried sending the passwords on server connect instead of NickServ? How about CertFP?

> If we could drop the +R from the #tor-meeting channel that would be ideal, but otherwise we will have to debug it more...

I would rather avoid that. Dropping that will just bring more spam in the channel and I would rather see this bot fixed than have to deal with spam on our end.

Looking at this code:

```
        "messageSplit": 448,
        "autoSendCommands": [
          ["PRIVMSG", "NickServ", "GHOST slacktopus {{ slacktopus_password }}" ],
          ["NICK", "slacktopus"],
          ["PRIVMSG", "NickServ", "IDENTIFY {{ slacktopus_password }}"]
        ],
```

... those first two line seem strange. Why do you need to send a `GHOST` command there and then *reset* the nick? That shouldn't be necessary unless multiple instances of the bot join at once. Worst case if the nick is in use for some obscure reason, it should fallback to another one and you can "group" the nick together so they can all be used at once. I rarely need to use the GHOST command and would really only use it if someone was really trying to mess with me....

I'll also note that the `slacktopus` nick is currently online on OFTC, but only joined to `#kandoo-iran`, which is weird because the other channels *should* also work, at least `#ooni`, because it's not `+R`.

In other words, TL;DR: your bot is buggy. ;)

**Comment by hellais** [comment]
> Have you tried sending the passwords on server connect instead of NickServ? How about CertFP?

I'm not sure how to do that. I tried using `sasl` authentication, but that was showing the errors I pasted above.

The bot we are using is this: https://github.com/ekmartin/slack-irc, which relies on this library: https://github.com/Throne3d/node-irc.

The documentation for the supported options are: https://node-irc-upd.readthedocs.io/en/latest/API.html#client.

I tried doing some research to see if it supports SSL options for client certificates, but could not find any example or reference to this, so it may not work.

> I would rather avoid that. Dropping that will just bring more spam in the channel and I would rather see this bot fixed than have to deal with spam on our end.

Fair enough, that makes sense.

> those first two line seem strange. Why do you need to send a GHOST command there and then reset the nick?

We added those lines because in the past we were experiencing a lot of times the bot connection would timeout and re-join with the old one still being connected and to avoid a nick conflict we had to do that command.

It has been quite stable recently so maybe it is worth dropping it and seeing if it works.

> I'll also note that the slacktopus nick is currently online on OFTC, but only joined to #kandoo-iran, which is weird because the other channels should also work, at least #ooni, because it's not +R.

That may have been due to the fact I was doing some debugging on it, but it should now be connected fine to all channels:
```
05:35 [oftc] -!- slacktopus [~slacktopu@37.218.241.90]
05:35 [oftc] -!-  ircname  : OONI Slack relay
05:35 [oftc] -!-  channels : #ooni-entropy #ooni-dev #ooni #kandoo-iran
```

> In other words, TL;DR: your bot is buggy. ;)

Yes I agree it doesn't work as well as it could, though supporting bridging with IRC is kind of best effort for us as most people are on the slack channel anyways.

I am considering as an alternative option for people allergic to slack to setup a matrix bridge, which has an official integration: https://matrix.org/docs/projects/bridge/matrix-appservice-slack/.

Supporting the IRC bridge has been a huge headache and maybe we should just give up on it and rather offer a more modern solution to the problem.

**Comment by hellais** [comment]
So it turns out the fix was actually much easier. There was wrong nesting in the autoSendCommands: https://github.com/ooni/sysadmin/commit/838d547e22368642561112be0f6bda469b91af91.

There is still a chance for a race, but it appears to be working alright now.

I am closing this issue as fixed.

**Comment by anarcat** [comment]
> > Have you tried sending the passwords on server connect instead of NickServ? How about CertFP?

> I'm not sure how to do that. I tried using sasl authentication, but that was showing the errors I pasted above.

Looking at the API docs you sent, it seems to me the `password` field should be all that you need. :)

```
{
    userName: 'nodebot',
    realName: 'nodeJS IRC client',
    password: null,
    port: 6667,
[...]
}
```

... but it does seem like you have tried that in b05a32a39c43dc561911753a61cf720ad842533b... 

> I tried doing some research to see if it supports SSL options for client certificates, but could not find any example or reference to this, so it may not work.

It indeed seems like that implementation doesn't support client certificates for SSL connections.

> We added those lines because in the past we were experiencing a lot of times the bot connection would timeout and re-join with the old one still being connected and to avoid a nick conflict we had to do that command.

I see... normally, IRC clients have a "fallback nick" they use to work around that problem. That nick gets selected when the first one is taken, and then the two nicks (normal and fallback) are grouped in nickserv so they can both be used to authenticate. This bot seems to use a different mechanism (the `renick` thing) of trying to retake the nick after a delay...

> Supporting the IRC bridge has been a huge headache and maybe we should just give up on it and rather offer a more modern solution to the problem.

I can imagine! I'm sorry you're having all this trouble... writing IRC clients is surprisingly hard as the protocol is old and kind of weird, but once you have a solid library, it usually works pretty well. The problem happens when people think "oh I know IRC, this is a simple text protocol i'll do this myself" and don't realize all those little race conditions and problems that can happen. ;)

> So it turns out the fix was actually much easier. There was wrong nesting in the autoSendCommands: ooni/sysadmin@838d547.

Aaah... so it was just not sending the commands at all! weird... Maybe that's why b05a32a39c43dc561911753a61cf720ad842533b didn't work as well...

> There is still a chance for a race, but it appears to be working alright now.

Right. Happy to see it worked out for you, and good luck for everything else! :)
